### PR TITLE
feat: Made booker layout settings more user friendly

### PR DESCRIPTION
**Pull request by JeroenReumkens**
## What does this PR do?

Makes booker layout settings a bit more user friendly. These things have changed:

1. When you disable a layout as one of the enabled options, the button to mark it as default layout disables too.
2. If it was chosen as a default before you disabled that layout, it takes the next enabled layout and makes that default now. 
3. If you only enable one layout, the settings for the default layout aren't shown at all

Fixes CAL-1889 CAL-1893 CAL-1894 CAL-1895

https://github.com/calcom/cal.com/assets/2969573/2fdfbc65-c2fa-42a1-88ea-b3b4997646f4

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

- [ ] Enable and disable layouts, and ensure that the behaviour is clear. 
- [ ] Test that the enabled layouts are also still shown correctly on the booker page
- [ ] Test with a user that doesn't have any settings set yet, and note that by default it now only shows the month view.

### CODE REVIEW

File: apps/web/pages/settings/my-account/appearance.tsx [hunk 1]
```diff
@@ -99,7 +99,7 @@ const AppearanceView = () => {
     <Form
       form={formMethods}
       handleSubmit={(values) => {
-        const layoutError = validateBookerLayouts(values.defaultBookerLayouts || null);
+        const layoutError = validateBookerLayouts(values?.metadata?.defaultBookerLayouts || null);
```
Comment (JeroenReumkens) [inline apps/web/pages/settings/my-account/appearance.tsx:102]: This is an unrelated issue. I noticed that I didn't change the fieldname after I moved the layout settings to metadata, so client side validation wasn't working. It is now.

File: packages/features/bookings/Booker/Booker.tsx [hunk 2]
```diff
@@ -51,10 +49,7 @@ const BookerComponent = ({
     shallow
   );
   const extraDays = layout === BookerLayouts.COLUMN_VIEW ? (isTablet ? 2 : 4) : 0;
-  const bookerLayouts = event.data?.profile?.bookerLayouts || {
-    defaultLayout: BookerLayouts.MONTH_VIEW,
-    enabledLayouts: bookerLayoutOptions,
-  };
+  const bookerLayouts = event.data?.profile?.bookerLayouts || defaultBookerLayoutSettings;
```
Comment (JeroenReumkens) [inline packages/features/bookings/Booker/Booker.tsx:52]: Instead of repeating the default settings I've moved it into a reusable variable.

File: packages/features/settings/BookerLayoutSelector.tsx [hunk 3]
```diff
@@ -85,19 +85,28 @@ const BookerLayoutFields = ({ settings, onChange, showUserSettings }: BookerLayo
   // Converts the settings array into a boolean object, which can be used as form values.
   const toggleValues: BookerLayoutState = bookerLayoutOptions.reduce((layouts, layout) => {
     layouts[layout] = !shownSettings?.enabledLayouts
-      ? true
```
Comment (JeroenReumkens) [inline packages/features/settings/BookerLayoutSelector.tsx:88]: Before if no settings were set, all layouts would be enabled. Now we only enable the layouts that are enabled in the default settings (so only month view as of now).

File: packages/features/settings/BookerLayoutSelector.tsx [hunk 4]
```diff
@@ -85,19 +85,28 @@ const BookerLayoutFields = ({ settings, onChange, showUserSettings }: BookerLayo
...
       onChange({
-        enabledLayouts: Object.keys(toggleValues).filter((layout) => {
-          if (changedLayout === layout) return checked === true;
-          return toggleValues[layout as BookerLayouts] === true;
-        }) as BookerLayouts[],
-        defaultLayout,
+        enabledLayouts: newEnabledLayouts,
+        // If default layout is toggled off, we set the default layout to the first enabled layout
+        // if there's none enabled, we set it to month view.
+        defaultLayout: isDefaultLayoutToggledOff
```
Comment (JeroenReumkens) [inline packages/features/settings/BookerLayoutSelector.tsx:107]: Added behaviour that if you disable the layout that's currently marked as the default, that we switch to the next best candidate. Otherwise you would have a layout marked as default that's actually a disabled radio button.

File: packages/features/settings/BookerLayoutSelector.tsx [hunk 5]
```diff
@@ -162,7 +172,8 @@ const BookerLayoutFields = ({ settings, onChange, showUserSettings }: BookerLayo
           onValueChange={(layout: BookerLayouts) => onDefaultLayoutChange(layout)}>
           {bookerLayoutOptions.map((layout) => (
             <RadioGroup.Item
-              className="aria-checked:bg-emphasis hover:bg-subtle focus:bg-subtle w-full rounded-[4px] p-1 text-sm transition-colors"
+              className="aria-checked:bg-emphasis hover:[&:not(:disabled)]:bg-subtle focus:[&:not(:disabled)]:bg-subtle w-full rounded-[4px] p-1 text-sm transition-colors disabled:cursor-not-allowed disabled:opacity-40"
```
Comment (JeroenReumkens) [inline packages/features/settings/BookerLayoutSelector.tsx:175]: Style changes to not show hover effects for disabled buttons.

File: packages/trpc/server/routers/loggedInViewer/updateProfile.handler.ts [hunk 6]
```diff
@@ -32,6 +34,13 @@ export const updateProfileHandler = async ({ ctx, input }: UpdateProfileOptions)
   };
 
   let isPremiumUsername = false;
+
+  const layoutError = validateBookerLayouts(input?.metadata?.defaultBookerLayouts || null);
+  if (layoutError) {
```
Comment (JeroenReumkens) [inline packages/trpc/server/routers/loggedInViewer/updateProfile.handler.ts:39]: There was now backend validation yet for the booker layouts in the update profile handler. There is now.

File: packages/features/settings/BookerLayoutSelector.tsx [hunk 7]
```diff
@@ -162,7 +172,8 @@ const BookerLayoutFields = ({ settings, onChange, showUserSettings }: BookerLayo
...
+              disabled={toggleValues[layout] === false}
```
Comment (JeroenReumkens) [inline packages/features/settings/BookerLayoutSelector.tsx:176]: This part disables this layout for being a default layout, if the layout itself is not enabled.

**Comment by PeerRich** [comment]
> Finally an important change: After discussing with @Jaibles we concluded that it might not be the best experience if ALL users suddenly have all layouts enabled on their booking pages after release. So instead of enabling all layouts by default when you have no setting, we actually only show the month view like in the old booker, until you manually override settings either in the event settings or in your user settings.

this has been undone right?

ideally everyone has all 3 layouts and its opt-out to remove them

**Comment by Jaibles** [comment]
> > Finally an important change: After discussing with @Jaibles we concluded that it might not be the best experience if ALL users suddenly have all layouts enabled on their booking pages after release. So instead of enabling all layouts by default when you have no setting, we actually only show the month view like in the old booker, until you manually override settings either in the event settings or in your user settings.
> 
> this has been undone right?
> 
> ideally everyone has all 3 layouts and its opt-out to remove them

as far as I know, yes.

**Comment by JeroenReumkens** [comment]
> > Finally an important change: After discussing with @Jaibles we concluded that it might not be the best experience if ALL users suddenly have all layouts enabled on their booking pages after release. So instead of enabling all layouts by default when you have no setting, we actually only show the month view like in the old booker, until you manually override settings either in the event settings or in your user settings.
> 
> this has been undone right?
> 
> ideally everyone has all 3 layouts and its opt-out to remove them

Correct. This isn't in there anymore. Let me update the description.

File: packages/trpc/server/routers/loggedInViewer/updateProfile.handler.ts [hunk 8]
```diff
@@ -32,6 +34,13 @@ export const updateProfileHandler = async ({ ctx, input }: UpdateProfileOptions)
...
+    const t = await getTranslation("en", "common");
```
Comment (Udit-takkar) [inline packages/trpc/server/routers/loggedInViewer/updateProfile.handler.ts:40]: why not  ` const t = await getTranslation(ctx.user.locale, "common");`
 ?

**Review by Udit-takkar — APPROVED** [review]
@JeroenReumkens LGTM great work. Left just one comment

**Comment by Jaibles** [comment]
@Udit-takkar is this good to merge or does your comment need addressing? If so can we jump in for @JeroenReumkens to address?

**Comment by Udit-takkar** [comment]
@Jaibles It is good to merge. he can address the comment later
